### DISCUSSION TITLE
Lotka–Volterra equations

### DISCUSSION THREAD

**Discussion by beantsja** [usnistgov/fipy#989] [Q&A] [answered]
How would I write an equation for Lotka-Volterra equation? It is because I am planning to expand this to writing a four coupled PDE that has a source term similar in this form. For example, I am solving for x,y,z,g, and the source terms are like x*y, y*z and etc...

<img width="169" alt="image" src="https://github.com/usnistgov/fipy/assets/155326827/f0da9755-2f6e-40cc-afb3-0e14e5c24baa">

```
eqn0 = TransientTerm(var=x) == a*ImplicitSourceTerm(coeff=1.0, var=x)-b*ImplicitSourceTerm(coeff=1.0, var=x)*ImplicitSourceTerm(coeff=1.0, var=y)
eqn1 = TransientTerm(var=y) == d*ImplicitSourceTerm(coeff=1.0, var=x)*ImplicitSourceTerm(coeff=1.0, var=y)-c*ImplicitSourceTerm(coeff=1.0, var=y)

eqn = eqn0 & eqn1
```

This is what I have, but it is giving me an error called "TermMultiplyError: Must multiply terms by int or float."

**Comment by guyer** [comment]
You cannot multiply a `Term` by another `Term`.

FiPy uses the Finite Volume method to discretize PDEs into linear algebra equations. Intrinsic to *linear* algebra is that `Term`s can only be implicit in one variable, so you must choose.

I would write
```python
eqn0 = TransientTerm(var=x) == ImplicitSourceTerm(coeff=a, var=x) - ImplicitSourceTerm(coeff=b * x, var=y)
eqn1 = TransientTerm(var=y) == ImplicitSourceTerm(coeff=d * y, var=x) - ImplicitSourceTerm(coeff=c, var=y)

eqn = eqn0 & eqn1
```

(The alternative,
```python
eqn0 = TransientTerm(var=x) == ImplicitSourceTerm(coeff=a, var=x) - ImplicitSourceTerm(coeff=b * y, var=x)
eqn1 = TransientTerm(var=y) == ImplicitSourceTerm(coeff=d * x, var=y) - ImplicitSourceTerm(coeff=c, var=y)

eqn = eqn0 & eqn1
```
would result in two equations that aren't coupled at all. While they would probably solve, they would probably converge more poorly.)

**Comment by guyer** [comment]
Beyond that, what you have so far are ODEs rather than PDEs. I'd think that FiPy could solve them, but tools like [`scipy.integrate.solve_ivp()`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.integrate.solve_ivp.html) or [SUNDIALS](https://scikits-odes.readthedocs.io/en/stable/solvers.html) would probably do better.

**Reply by beantsja** [reply]
![image](https://github.com/usnistgov/fipy/assets/155326827/acef498a-4d5a-4e89-8a2e-9607d8249eb1)
<img width="434" alt="image" src="https://github.com/usnistgov/fipy/assets/155326827/aea4a6db-b671-4e7f-8c96-3b7f0218c6dc">
![image](https://github.com/usnistgov/fipy/assets/155326827/c63cced4-40d3-4d95-92c1-1c3034276a1d)

Your suggested equations seem to work!

But I have one concern. 

The first picture is what I have, and the second picture is what I am supposed to have. Third picture is with smaller dt. I used the first equations that you provided. When I use smaller dt, the graph becomes totally linear. What seems to be the issue or is there a limit on dt? 

This is my code in case it helps. 

```
from fipy import Grid1D, CellVariable, TransientTerm, DiffusionTerm, Viewer, ImplicitSourceTerm
import matplotlib.pyplot as plt
import numpy as np


m = Grid1D(nx=100, Lx=1.)

x = CellVariable(mesh=m, hasOld=True, value=80.0)
y = CellVariable(mesh=m, hasOld=True, value=20)


# x.constrain(0, m.facesLeft)
# x.constrain(0, m.facesRight)

# y.constrain(0, m.facesLeft)
# y.constrain(0, m.facesRight)

a = 0.1
b = 0.002
c = 0.2
d = 0.0025

vi = Viewer((x, y))


eqn0 = TransientTerm(var=x) == ImplicitSourceTerm(coeff=a, var=x) - ImplicitSourceTerm(coeff=b * x, var=y)
eqn1 = TransientTerm(var=y) == ImplicitSourceTerm(coeff=d * y, var=x) - ImplicitSourceTerm(coeff=c, var=y)

eqn = eqn0 & eqn1



# Set up the Matplotlib plot
plt.ion()  # Interactive mode on
plt.figure()
# Get the x-coordinates from the mesh
x_coords = m.cellCenters[0]
x_values_over_time = []
y_values_over_time = []

time = 100

from builtins import range
for t in range(time):
    # Update the plot
    x.updateOld()
    y.updateOld()
    eqn.solve(dt=1)
    # Store the average value of x
    x_values_over_time.append(np.mean(x.value))
    y_values_over_time.append(np.mean(y.value))

    
# Turn off interactive plotting
plt.ioff()

# Plot x values over time
print(x_values_over_time)
plt.figure()
plt.plot(range(time), (x_values_over_time))
plt.plot(range(time), (y_values_over_time)) 
plt.xlabel('Time')
plt.ylabel('Value of x')
plt.title('Evolution of x Over Time')
plt.show()    
```
    
    
    
    
    
    
    


I think it's better to give my whole context, so it may help. After learning how to multiply the source terms that is also the variable that we solve, I wanted to apply in my situation here like this. This is the two coupled diffusion PDE solving for a and b. As you can see, a and b are multiplied together as a source term(where k is a constant). Can FiPy handle this type of situation?  How would FiPy solve them, if it is possible?  I am planning to have four coupled diffusion PDE just like this. 

Also, could you briefly explain how you arrive your first suggested equations? You put x as a coefficient in one of the ImplicitSourceTerm and y for another. Does it matter which variable I choose to put as a coefficient? In four coupled PDE case, can I put more than one variables in the coefficient? For example, 

`eqn0 = TransientTerm(var=x) == ImplicitSourceTerm(coeff=a, var=x) - ImplicitSourceTerm(coeff=b * x*z, var=y)
`
considering it is solving for x, y, z. 

<img width="163" alt="image" src="https://github.com/usnistgov/fipy/assets/155326827/5f37c058-030b-4041-9e5a-239726775a51">

**Reply by guyer** [reply]
There are a few things going on here:
- Your equations are nonlinear, so they require [sweeping](https://pages.nist.gov/fipy/en/latest/FAQ.html#iterations-timesteps-and-sweeps-oh-my) to reach convergence. I find that three sweeps is sufficient to reach convergence for your equations.
    ```diff
    48c48,50
    <     eqn.solve(dt=1)
    ---
    >     for sweep in range(3):
    >         res = eqn.sweep(dt=1)
   ```
- By setting the initial value to `20`, you're storing integers in `y`. There are reasons to do that, but integers don't work well for solution variables. I'm not sure why you're not getting a warning about that; perhaps because you've coupled with `x` containing floats.
    ```diff
    9c9
    < y = CellVariable(mesh=m, hasOld=True, value=20)
    ---
    > y = CellVariable(mesh=m, hasOld=True, value=20.)
    ```
- With smaller timesteps, the solutions are only (apparently) linear over the same hundred steps. If you, e.g., increase to 1000 steps of 0.1 (for the same total time), then you get fluctuating solutions. With `dt=1`, I find that the amplitude of `y` damps over time.
    ![step by dt=1](https://github.com/usnistgov/fipy/assets/1830725/eb0d4ab7-f4c7-47f3-9ba0-97a1b0c28cf2)
    Decreasing the timestep to `dt=0.1`
    ![step by dt=0.1](https://github.com/usnistgov/fipy/assets/1830725/fcd8bebf-b80c-4078-ab50-23541520209e)
    or `dt=0.01`
    ![step by dt=0.01](https://github.com/usnistgov/fipy/assets/1830725/3861c54c-f670-4835-8fbf-d880c12374d6)
    leads to solutions with more consistent amplitude.
    Smaller timesteps do improve accuracy, but they are quite slow. This is both because FiPy isn't designed for solving ODEs and because you're solving the ODE 100 times (at every grid point).

**Reply by guyer** [reply]
Your planned equations

$$
\begin{aligned}
\frac{\partial a}{\partial t} &= D_a \frac{\partial^2 a}{\partial x^2} + k b a
\\
\frac{\partial b}{\partial t} &= D_b \frac{\partial^2 b}{\partial x^2} - k b a
\end{aligned}
$$

are definitely more suitable for FiPy. These would be written as
```python
eqn0 = TransientTerm(var=a) == DiffusionTerm(coeff=Da, var=a) + ImplicitSourceTerm(coeff=k * a, var=b)
eqn1 = TransientTerm(var=b) == DiffusionTerm(coeff=Db, var=b) - ImplicitSourceTerm(coeff=k * b, var=a)
```

**Reply by beantsja** [reply]
Thank you very much. 

Lastly, could you explain how you derive the equations?

```
eqn0 = TransientTerm(var=x) == ImplicitSourceTerm(coeff=a, var=x) - ImplicitSourceTerm(coeff=b * x, var=y)
eqn1 = TransientTerm(var=y) == ImplicitSourceTerm(coeff=d * y, var=x) - ImplicitSourceTerm(coeff=c, var=y)

eqn = eqn0 & eqn1
```
You put x as a coefficient in one of the ImplicitSourceTerm and y for another. Does it matter which variable I choose to put as a coefficient? In four coupled PDE case, can I put more than one variables in the coefficient? For example,

`eqn0 = TransientTerm(var=x) == ImplicitSourceTerm(coeff=a, var=x) - ImplicitSourceTerm(coeff=b * x*z, var=y) `
considering it is solving for x, y, z.

**Reply by guyer** [reply, marked as answer]
You can write it either way, but it matters. `ImplicitSourceTerm(coeff=f, var=g)` represents $f g$. `f` can be anything, but the expression can only be implicit (linear) in one variable, `g`. 

By choosing
```python
eqn0 = TransientTerm(var=x) == ImplicitSourceTerm(coeff=a, var=x) - ImplicitSourceTerm(coeff=b * x, var=y)
eqn1 = TransientTerm(var=y) == ImplicitSourceTerm(coeff=d * y, var=x) - ImplicitSourceTerm(coeff=c, var=y)
```
I chose to represent

$$
\begin{align}
x' &= a x - (b x) y
\\
y' &= (d y) x - c y
\end{align}
$$

or

$$
\begin{align}
(\partial_t - a) x - (b x) y &= 0
\\
 -(d y) x + (\partial_t - c) y &= 0
\end{align}
$$

which creates a (***very*** simplified) block matrix

$$
\left[\begin{array}{c|c} 
	\partial_t - a & -b x \\ 
	\hline 
	-d y & \partial_t - c 
\end{array}\right] 
\left[\begin{array}{c}
x \\
y
\end{array}\right]
= 0
$$

which represents pretty good coupling.





If you write
```python
eqn0 = TransientTerm(var=x) == ImplicitSourceTerm(coeff=a, var=x) - ImplicitSourceTerm(coeff=b * y, var=x)
eqn1 = TransientTerm(var=y) == ImplicitSourceTerm(coeff=d * x, var=y) - ImplicitSourceTerm(coeff=c, var=y)
```
then you are saying

$$
\begin{align}
x' &= a x - (b y) x
\\
y' &= (d x) y - c y
\end{align}
$$

or

$$
\begin{align}
(\partial_t - a - b y) x &= 0
\\
(\partial_t - d x - c) y &= 0
\end{align}
$$

which creates a block matrix

$$
\left[\begin{array}{c|c} 
	\partial_t - a - b y & 0 \\ 
	\hline 
	0 & \partial_t - d x - c
\end{array}\right] 
\left[\begin{array}{c}
x \\
y
\end{array}\right]
= 0
$$

The first equation solves only for `x` and the second solves only for `y`. You can "couple" them with `&`, but they're not really coupled and they won't converge any better than if you swept them sequentially. It'll still work, but it will probably converge more slowly. 

Actually, it looks like the convergence is about the same either way, with the residual reaching machine precision in around 20 sweeps. I'd originally identified 3 sweeps convergence when `y` was still an integer. To be honest, the solution doesn't look any "better" after 20 sweeps. Accuracy (assuming constant amplitude oscillations is "accurate") is only achieved by reducing the time step.

**Reply by guyer** [reply]
Thinking about this a little more, the fact that the off-diagonal blocks ($-bx$ and $-dy$) are negative probably means that they get put on the right-hand-side vector instead of on the matrix (for stability reasons). This is probably why the two different approaches have similar convergence. Still, I'd do the first one; the second is definitely not coupled, whereas FiPy will try to couple the first if it can.
